apm-agent-core/src/main/java/co/elastic/apm/bci/ElasticApmAgent.java

```diff
@@ -122,44 +139,80 @@ private static AgentBuilder applyAdvice(final ElasticApmTracer tracer, final Age
         final Logger logger = LoggerFactory.getLogger(ElasticApmAgent.class);
         logger.debug("Applying advice {}", advice.getClass().getName());
         advice.init(tracer);
+        final boolean typeMatchingWithNamePreFilter = tracer.getConfig(CoreConfiguration.class).isTypeMatchingWithNamePreFilter();
         return agentBuilder
             .type(new AgentBuilder.RawMatcher() {
                 @Override
                 public boolean matches(TypeDescription typeDescription, ClassLoader classLoader, JavaModule module, Class<?> classBeingRedefined, ProtectionDomain protectionDomain) {
-                    boolean typeMatches;
+                    long start = System.nanoTime();
                     try {
-                        typeMatches = advice.getTypeMatcher().matches(typeDescription);
-                    } catch (Exception ignored) {
-                        // happens for example on WebSphere, not sure why ¯\_(ツ)_/¯
-                        typeMatches = false;
-                    }
-                    if (typeMatches) {
-                        logger.debug("Type match for advice {}: {} matches {}",
-                            advice.getClass().getSimpleName(), advice.getTypeMatcher(), typeDescription);
-                        if (logger.isTraceEnabled()) {
-                            logClassLoaderHierarchy(classLoader, logger, advice);
+                        if (typeMatchingWithNamePreFilter && !advice.getTypeMatcherPreFilter().matches(typeDescription)) {
+                            return false;
+                        }
+                        boolean typeMatches;
+                        try {
+                            typeMatches = advice.getTypeMatcher().matches(typeDescription);
```

Comment:
I'd rather leave it as-is. That way I capture the total overhead of the `AgentBuilder.RawMatcher#matches` method. Also, the time measurement is better separated from the actual logic. It makes it easier to add logic, without thinking about where to set the measurement code (in another PR I'm adding matching based on the classloader in this method) and it would also make it easier to extract the measurement code in another class.

Reply:
Well, if you really use if for debug, the debug will be on, so your timing will include actions related to the logging actions...